build-logic/src/main/java/qsl/internal/json/QmjBuilder.java

```diff
@@ -0,0 +1,119 @@
+				.endObject()
+				.beginObject()
+				.name("id").value("minecraft")
+					.name("versions").value(minecraftVersion)
+					.endObject();
+		for (QslLibraryDependency depend : ext.getModuleDependencyDefinitions()) {
+			for (QslLibraryDependency.ModuleDependencyInfo moduleDependencyInfo : depend.getDependencyInfo().get()) {
+				Project depProject = project.getRootProject().project(depend.getName()).project(moduleDependencyInfo.module());
+				QslModuleExtension depExt = depProject.getExtensions().getByType(QslModuleExtension.class);
+				writer.beginObject()
+						.name("id").value(depExt.getId().get())
+						.name("versions").value(depProject.getVersion().toString())
+						.endObject();
+			}
+		}
+		writer.endArray(); // depends -> quilt_loader
+
+		if (!ext.getEntrypoints().isEmpty()) {
+			writer.name("entrypoints").beginObject(); // quilt_loader -> entrypoints
+			for (QslModuleExtensionImpl.NamedWriteOnlyList entrypoint : ext.getEntrypoints()) {
+				writer.name(entrypoint.getName());
+				writer.beginArray();
+				for (String clazz : entrypoint.getValues().get()) {
+					writer.value(clazz);
+				}
+				writer.endArray();
+			}
+			writer.endObject(); // entrypoints -> quilt_loader
+		}
+		writer.endObject(); // quilt_loader -> root object
+
+		if (ext.getHasMixins().get()) {
+			writer.name("mixin").value(ext.getId().get() + ".mixins.json");
+		}
+
+		if (ext.getHasAccessWidener().get()) {
+			writer.name("access_widener").value(ext.getId().get() + ".accesswidener");
+		}
+
+		// TODO: environment
+		if (ext.getEnvironment().get() != Environment.ANY) {
+			writer.name("minecraft").beginObject() // root object -> minecraft
+					.name("environment").value(ext.getEnvironment().get().qmj)
+					.endObject(); // minecraft -> root object
+		}
+		writer.name("modmenu").beginObject() // root object -> modmenu
+				.name("badges").beginArray().value("library").endArray()
+				.name("parent").beginObject() // modmenu -> parent
+				.name("id").value("qsl")
+				.name("name").value("Quilt Standard Libraries")
+				.name("description").value("A set of libraries to assist in making Quilt mods")
+				.name("badges").beginArray().value("library").endArray()
+				.endObject() // parent -> modmenu
+				.endObject(); // modmenu -> root
+
+		if (!ext.getInjectedInterfaces().isEmpty()) {
+			writer.name("quilt_loom").beginObject() // root object -> quilt_loom
+					.name("injected_interfaces").beginObject(); // quilt_loom -> injected_interfaces
+
+			for (QslModuleExtensionImpl.NamedWriteOnlyList injectedInterface : ext.getInjectedInterfaces()) {
+				writer.name(injectedInterface.getName());
+				writer.beginArray();
+
+				for (String clazz : injectedInterface.getValues().get()) {
+					writer.value(clazz);
+				}
+
+				writer.endArray();
+			}
+
+			writer.endObject() // injected_interfaces -> quilt_loom
+					.endObject(); // quilt_loom -> root object
+		}
+
+		writer.endObject(); // end root object
+		writer.flush();
+		writer.close();
```

Comment:
can we make sure that the indentation here always indents for `beginObject`

Reply:
No. Too much of a pain in the ass to do in IDEA.

Reply:
If you try to indent further it will snap back to 1 level of indentation. I don't particularly feel like figuring out why...